File: Drakborgen/Vault.cpp
#include "Vault.h"

#include "System.h"

std::unique_ptr<Vault> Vault::makeVault(Direction section)
{
	switch (section)
	{
	case Direction::North:
		return std::make_unique<Vault>(
			std::vector<Direction>({ Direction::North }),
			std::vector<Direction>({ Direction::East }),
			std::vector<Direction>({ Direction::West }),
			std::vector<Direction>({ Direction::South, Direction::East }),
			std::vector<Direction>({ Direction::South, Direction::South }),
			std::vector<Direction>({ Direction::South, Direction::West })
			);
	case Direction::South:
		return std::make_unique<Vault>(
			std::vector<Direction>({ Direction::South }),
			std::vector<Direction>({ Direction::East }),
			std::vector<Direction>({ Direction::West }),
			std::vector<Direction>({ Direction::North, Direction::East }),
			std::vector<Direction>({ Direction::North, Direction::North }),
			std::vector<Direction>({ Direction::North, Direction::West })
			);
	default:
		THROW;
	}
}

std::vector<std::vector<Direction>> Vault::getExits(Direction) const
{
	return exits;
}

File: DrakborgenDotNet/Vault.cs

namespace Drakborgen
{
    internal class Vault : Tile
    {
        private List<List<Direction>> _exits;

        internal Vault(List<List<Direction>> exits)
            : base(null)
        {
            _exits = exits;
        }

        internal static Tile MakeVault(Direction section)
        {
            switch (section)
            {
                case Direction.North:
                    return new Vault(
                        [
                            [Direction.North],
                            [Direction.East],
                            [Direction.West],
                            [Direction.South, Direction.East],
                            [Direction.South, Direction.South],
                            [Direction.South, Direction.West]
                        ]
                    );
                case Direction.South:
                    return new Vault(
                        [
                            [Direction.South],
                            [Direction.East],
                            [Direction.West],
                            [Direction.North, Direction.East],
                            [Direction.North, Direction.North],
                            [Direction.North, Direction.West]
                        ]
                    );
                default:
                    throw new Exception();
            }
        }

        internal override List<List<Direction>> GetExits(Direction direction)
        {
            return _exits;
        }

        internal override State Enter()
        {
            Console.WriteLine(" -> turn end");
            return State.TurnEnd;
        }
    }
}